### ISSUE TITLE
Impl Read<u8> / Write<u8> for Serial

### ISSUE THREAD

**Issue by dbrgn** [stm32-rs/stm32f1xx-hal#167]
Is there a reason why Read<u8> / Write<u8> is only implemented for `Rx` and `Tx` (obtainable via `split()`) but not on the `Serial` type directly?

Other HALs like the [stm32f4xx-hal](https://docs.rs/stm32f4xx-hal/0.6.0/stm32f4xx_hal/serial/struct.Serial.html) have such impls.

I'm working with a driver that wants a `Read<u8> + Write<u8>` type and I don't want to create a wrapper type with my own impl if not necessary :slightly_smiling_face:

**Comment by TheZoq2** [comment]
I don't believe there is an underlying reason, it's just that nobody has gotten around do doing it yet  :)

**Comment by dbrgn** [comment]
Hm, I tried to implement it, but I realized that for this struct:

```
pub struct Serial<USART, PINS> {
    usart: USART,
    pins: PINS,
}
```

...the Read / Write traits cannot be implemented, as it is not known how to get to the pin types from the generic `PINS` type argument.

One workaround would be to implement `Read / Write` for a tuple `(Tx, Rx)`, but that's a bit of a hack in my opinion.

The better solution in my eyes would be to replace the `Pins` trait with separate `TxPin` / `RxPin` traits, as I did in the [stm32l0xx-hal](https://github.com/stm32-rs/stm32l0xx-hal/pull/68). Any opinions on that? It would be a breaking change, but we're still on version 0.x so I guess that's fine :slightly_smiling_face:
